tests/unit/test_charm.py

```diff
@@ -498,50 +498,40 @@ def test_given_can_connect_when_install_then_ca_certificate_pushed_to_workload(
             ]
         )
 
-    @patch("vault.Vault.unseal", new=Mock)
+    @patch("vault.Vault.unseal")
     @patch("vault.Vault.initialize")
     @patch("vault.Vault.is_api_available")
-    @patch("ops.model.Container.exists")
-    @patch("ops.model.Container.pull")
-    @patch("ops.model.Container.push")
-    @patch("ops.model.Model.get_binding")
-    @patch("charm.generate_vault_unit_certificate")
-    @patch("charm.generate_vault_ca_certificate")
     def test_given_certificates_pushed_when_install_then_certificate_not_pushed_to_workload(
         self,
-        patch_generate_ca_certs,
-        patch_generate_unit_certs,
-        patch_get_binding,
-        patch_push,
-        patch_pull,
-        patch_exists,
         patch_is_api_available,
         patch_vault_initialize,
+        patch_vault_unseal,
     ):
-        ca_private_key = "ca private key content"
-        ca_certificate = "ca certificate content"
-        unit_private_key = "unit private key content"
-        unit_certificate = "unit certificate content"
-
-        patch_exists.return_value = True
-        patch_pull.return_value = StringIO()
-        patch_generate_ca_certs.return_value = ca_private_key, ca_certificate
-        patch_generate_unit_certs.return_value = unit_private_key, unit_certificate
-        bind_address = "1.2.3.4"
-        ingress_address = "10.1.0.1"
-        patch_get_binding.return_value = MockBinding(
-            bind_address=bind_address, ingress_address=ingress_address
-        )
         patch_is_api_available.return_value = True
         patch_vault_initialize.return_value = "root token content", "unseal key content"
+
+        # Pre-create dummy CA and key+cert so install doesn't create them
+        root = self.harness.get_filesystem_root(self.container_name)
+        (root / "vault/config").mkdir(parents=True)
+        (root / "vault/certs").mkdir(parents=True)
+        (root / "vault/certs/ca.pem").write_text("CA")
+        (root / "vault/certs/key.pem").write_text("KEY")
+        (root / "vault/certs/cert.pem").write_text("CERT")
+
+        self.harness.add_network("1.2.3.4")
         self._set_peer_relation()
         self.harness.set_leader(is_leader=True)
         self.harness.set_can_connect(container=self.container_name, val=True)
 
         self.harness.charm.on.install.emit()
 
-        for call_args in patch_push.call_args_list:
-            self.assertFalse(call_args.kwargs["path"].startswith("/vault/certs"))
+        # Ensure CA and key+cert weren't overwritten
+        self.assertEqual((root / "vault/certs/ca.pem").read_text(), "CA")
+        self.assertEqual((root / "vault/certs/key.pem").read_text(), "KEY")
+        self.assertEqual((root / "vault/certs/cert.pem").read_text(), "CERT")
+
+        # Ensure vault was actually unsealed
+        patch_vault_unseal.assert_called_once()
```

Comment:
I'm not sure why we added this here, that's not what the test name claims to validate

Reply:
See my comment in the PR description:

> Also assert that vault.unseal() was actually called. It was too easy to break the test and accidentally follow the wrong code path, and the test still succeeded. For example, if you removed the patch_get_binding before, the test would still pass fine, but would be following one of the early returns -- not what the test was trying to test for.

I think it sort of is what the test is trying to validate (previously the charm event could return early, not follow the code path we were looking for at all, but still pass. We're checking that certificates are *not* pushed (something did *not* happen), so we have to validate that it actually got up to the certificate check, rather than returned before then.

This might not be the right assertion though. Feel free to suggest something else.

Could you guys please take this from here, as you know the charm a lot better? (And can resolve the conflicts. :-)

Reply:
Yes, I've created a new PR to apply this proposal all over the place:
- https://github.com/canonical/vault-k8s-operator/pull/59

Reply:
Excellent, thanks!